### PR TITLE
[TT-12355] added log format config, func and test cases

### PR DESCRIPTION
**Pull request by LLe27**
### **FR Jira Ticket**

https://tyktech.atlassian.net/browse/TT-12355

## Description

- Added the `TYK_GW_LOGFORMAT` Gateway config option
- The Tyk Gateway will determine which format should be used for STDOUT logs
  - If the `TYK_GW_LOGFORMAT` is set to `json` the Gateway will output logs in JSON format
  - if the `TYK_GW_LOGFORMAT` is set to `""` empty string or any other string value it will default the TEXT format 

Additionally, TYK_LOGFORMAT is supported, as is TYK_LOGLEVEL (and TYK_GW_LOGLEVEL).

Note that there is some global use from the log package, I don't think we can set json output from json config, as it would be picked up too late. The one case to cover is to make sure this passes with only config enabled:

```
# TYK_GW_LOGFORMAT=json ./tyk 2>&1 | head -n 1
{"level":"info","msg":"Tyk API Gateway v5.3.0-dev","prefix":"main","time":"2024-07-10T15:16:11+02:00"}
# TYK_LOGFORMAT=json ./tyk 2>&1 | head -n 1
{"level":"info","msg":"Tyk API Gateway v5.3.0-dev","prefix":"main","time":"2024-07-10T15:16:14+02:00"}
# ./tyk 2>&1 | head -n 1
{"level":"info","msg":"Tyk API Gateway v5.3.0-dev","prefix":"main","time":"2024-07-10T15:45:02+02:00"}
```

The above confirms reading from env and from config.

## Motivation and Context
The option to determine which STDOUT log format will open the opportunity for customers to easily parse the STDOUT logs and open up opportunity for our team to easily write tests and parse as needed.

## How This Has Been Tested

- Manual testing
- Unit testing 
- Benchmarks added
___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added the `TYK_GW_LOGFORMAT` Gateway config option to determine log format.
- Implemented `NewFormatter` function to switch between JSON and Text log formats.
- Modified log initialization to use the new formatter function.
- Added unit tests to verify the correct formatter is used based on the input format.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>log.go</strong><dd><code>Add log format configuration and initialization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

log/log.go
<li>Added <code>NewFormatter</code> function to handle log format based on <br><code>TYK_GW_LOGFORMAT</code> environment variable.<br> <li> Modified the default log formatter initialization to use <code>NewFormatter</code>.<br> <br>

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6344/files#diff-894630a7ed925c7768a861c4465bc9ad393f7937df7cac5c77ab123c03921aee">+17/-6</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>log_test.go</strong><dd><code>Add unit tests for log formatter configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

log/log_test.go
<li>Added unit tests for <code>NewFormatter</code> function.<br> <li> Tested JSON and Text formatters based on input format.<br>

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6344/files#diff-86c6c547a3e3c686a598e339acecbf5190185d424d96f9b9a0fff3d8513a90eb">+61/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: log/log_test.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package log
+
+import (
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"testing"
+	"time"
+)
+
+func TestNewFormatter(t *testing.T) {
```
Comment (titpetric) [inline log/log_test.go:10]: See the trouble with this is that it's ai generated :D you need to add some additional prompts.

You need to call NewFormatter with "" (any), and a "json", assert that the expected type is returned. (`_, ok := result.(*T) ; assert.True(t, ok)`). SLOC 4, coverage 100%.

File: gateway/server.go [hunk 2]
```diff
@@ -1245,6 +1245,18 @@ func (gw *Gateway) initialiseSystem() error {
 	overrideTykErrors(gw)
 
 	gwConfig := gw.GetConfig()
+	// Initialize the appropriate log formatter
+	if gwConfig.LogFormat != "" {
+		format := strings.ToLower(gwConfig.LogFormat)
+		log.Formatter = logger.NewFormatter(gwConfig.LogFormat)
+		switch format {
```
Comment (titpetric) [inline gateway/server.go:1252]: Most of this is redundant:

- dont use strings.ToLower, it should accept 'json', not 'JSON', etc.
- all of this needs to be something like 1 lines:

```
log.Formatter = logger.NewFormatter(gwConfig.LogFormat)
```

If you want to print to the log, do it once and pass %s with a sanitized LogFormat; ("standard", "json"), using log.

You're using `log` here (global), and `mainLog`? Are these global loggers all covered by setting the formatter just on `log`?

File: gateway/server.go [hunk 3]
```diff
@@ -1245,6 +1245,9 @@ func (gw *Gateway) initialiseSystem() error {
 	overrideTykErrors(gw)
 
 	gwConfig := gw.GetConfig()
+	// Initialize the appropriate log formatter
+	log.Formatter = logger.NewFormatter(gwConfig.LogFormat)
```
Comment (titpetric) [inline gateway/server.go:1249]:
```suggestion
	if gwConfig.LogFormat != "" {
		log.Formatter = logger.NewFormatter(gwConfig.LogFormat)
	}
```

Comment (titpetric) [reply gateway/server.go:1249]: This should keep the formatter from resetting whatever value was configured with the `TYK_LOGFORMAT` env (testing etc.).

**Comment by titpetric** [comment]
I have covered the change with some benchmarks.

```
# go test -count=1 -bench=. -benchtime=10s -benchmem ./...
goos: linux
goarch: amd64
pkg: github.com/TykTechnologies/tyk/log
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkFormatter/json-8         	 3902948	      3060 ns/op	    2483 B/op	      34 allocs/op
BenchmarkFormatter/json-ext-8     	 4970361	      2469 ns/op	    2076 B/op	      22 allocs/op
BenchmarkFormatter/default-8      	 4137858	      2715 ns/op	    1793 B/op	      27 allocs/op
BenchmarkFormatter/none-8         	11667763	      1062 ns/op	    1457 B/op	      12 allocs/op
PASS
ok  	github.com/TykTechnologies/tyk/log	57.376s
```

This shows improved performance under `json-ext` over both the json and the default text formatting, increasing throughput from ~390K (json), ~440K (standard) to ~528K logging calls per second.

**Review by andrei-tyk — APPROVED** [review]

**Pull request merged**